Code analysis for Laravel 9 branch

bennothommo opened the pull request:
Thought I would try installing Larastan/PHPStan and run some code analysis to look for some invalid PHP docblocks and dead code and the like. So far, it's picked up about 1100 errors - not sure if it's a bridge too far to cross, but interesting nonetheless.

We're testing against Level 5 so far, which seems to be a good compromise between ensuring all docblocks and signatures are correct whilst not having PHPStan being pedantic about *everything*.

## Modules

- [x] Argon
- [x] Auth
- [x] Config
- [x] Cookie
- [x] Database
- [x] Events
- [x] Exception
- [x] Extension
- [x] Filesystem
- [x] Flash
- [x] Foundation
- [x] Halcyon
- [x] Html
- [x] Http
- [x] Mail
- [x] Network
- [x] Parse
- [x] Redis
- [x] Router
- [x] Scaffold
- [x] Support
- [x] Translation
- [x] Validation

## Breaking changes introduced

We will need to discuss these before merging.

#### Database

- To prevent unsafe model instantiating, the model constructors are now forced to only allow a single `$attributes` parameter via an interface (`\Winter\Storm\Database\ModelInterface` and `\Winter\Storm\Halcyon\ModelInterface`). This will ensure that calls like `Model::make()` or `Model::create()` will execute correctly. It is possible that some people might have used additional parameters for their model constructors - these will no longer work and must be moved to another method.
- Due to the above, Pivot model construction has been rewritten. The constructor used to allow 4 parameters but now only allows the one `$attributes` parameter as per the Model class. Construction now happens more closely to Laravel's format of calling a `Pivot::fromAttributes()` static method. If you previously used `new Pivot()` to create a pivot model, switch to using `Pivot::fromAttributes()` instead.
- The `MorphToMany` class now extends the `MorphToMany` class from Laravel, as opposed to the `BelongsToMany` class in Winter. This prevents repeated code in the Winter `MorphToMany` class and maintains covariance with Laravel. This will mean that it will no longer inherit from the Winter `BelongsToMany` class. To allow for this, we have converted most of the overridden `BelongsToMany` functionality in Winter into a trait (`Concerns\BelongsOrMorphsToMany`). The `BelongsToMany` class now also uses this trait.
- The relation traits found in `src/Database/Relations` have been moved to `src/Database/Relations/Concerns`, in order to keep just the actual relation classes within this directory. In the unlikely event that you are using a relation trait directly, please rename the trait classes to the following:
    - `Winter\Storm\Database\Relations\AttachOneOrMany` to `Winter\Storm\Database\Relations\Concerns\AttachOneOrMany`
    - `Winter\Storm\Database\Relations\DeferOneOrMany` to `Winter\Storm\Database\Relations\Concerns\DeferOneOrMany`
    - `Winter\Storm\Database\Relations\DefinedConstraints` to `Winter\Storm\Database\Relations\Concerns\DefinedConstraints`
    - `Winter\Storm\Database\Relations\HasOneOrMany` to `Winter\Storm\Database\Relations\Concerns\HasOneOrMany`
    - `Winter\Storm\Database\Relations\MorphOneOrMany` to `Winter\Storm\Database\Relations\Concerns\MorphOneOrMany`

#### Extension

- Previously, the `Winter\Storm\Extension\Extendable::extendClassWith()` method returned the current class if the extension name provided was an empty string. This appears to be a code smell, so this has been changed to throw an Exception.

#### Filesystem

- The `Filesystem::symbolizePath` method's `$default` parameter now accepts a `string`, `bool` or `null`. Only in the case of `null` will the method return the original path - the given `$default` will be used in all other cases. This is the same as the original functionality, but we're documenting it in case you have customised this method in some fashion.
- The `Filesystem::isPathSymbol` method previously returned the path symbol used, as a string, if one was found and `false` if not found. However, the docblock stipulated, as well as the method name itself implied, that this is a boolean check function, so we have converted this to a straight boolean response - `true` if a path symbol is used, otherwise `false`.
- Many classes within the `Filesystem` namespace have had type hinting and return types added to enforce functionality and clarify documentation. If you extend any of these classes in your plugins, you may need to update your method signatures.

#### Foundation

- The `Application` class callback methods `before` and `after` were documented as `void` methods, but returned a value. These no longer return a value.

#### Halcyon

- To prevent unsafe model instantiating, the model constructors are now forced to only allow a single `$attributes` parameter via an interface (`\Winter\Storm\Database\ModelInterface` and `\Winter\Storm\Halcyon\ModelInterface`). This will ensure that calls like `Model::make()` or `Model::create()` will execute correctly. It is possible that some people might have used additional parameters for their model constructors - these will no longer work and must be moved to another method.
- The Halcyon Builder `insert` method now returns an `int` representing the created model's filesize, not a `bool`.
- The Halcyon Builder `insert` method requires the `$values` parameter to actually contain variables and will throw an Exception if an empty array is provided. Previously, this was silently discarded and returned `true` (although this would not actually save the file).

#### HTML

- The `Winter\Storm\Html\FormBuilder` class has had type hinting and return types added to enforce functionality and clarify documentation. If you extend this class in your plugin, you may need to make changes to your method signatures if you overwrite the base functionality.

#### Parse
- The `Bracket` class constructor is now `final` to prevent unsafe static calls to the `parse` static method.
- The `Bracket::parseString()` method previously returned `false` if a string was not provided for parsing, or the string was empty after trimming. Since the signature calls for a string, we won't check for this anymore. If the string is empty, an empty string will be returned.
- The `markdown.beforeParse` event in the `Markdown` class sent a single `MarkdownData` instance as a parameter to the listeners. It now sends an array with the `MarkdownData` instance as the only value, to meet the signature requirements for events.
- The `Syntax\FieldParser` class constructor has been made `final` to prevent unsafe static calls to the `parse` statuc method.
- The `$template` parameter for the `Syntax\FieldParser` class constructor was originally optional. Since there is no purpose for this, and no way to populate the template after the fact, this has been made required.
- The `Syntax\Parser` class constructor has been made `final` to prevent unsafe static calss to the `parse` static method.
- The `$template` parameter for the `Syntax\Parser` class constructor was originally optional. Since there is no purpose for this, and no way to populate the template after the fact, this has been made required.

#### Testing
- The `Winter\Storm\Support\Testing\Fakes\MailFake::queue()` method has had its signature re-arranged to remain compatible with Laravel's `MailFake` class, with the `$queue` parameter being moved to the second parameter. The new signature is `($view, $queue, $data, $callback)`.

bennothommo committed:
Add Larastan, initial code analysis tweaks

bennothommo committed:
Merge branch 'wip/0.2' into wip/1.2-code-analysis

bennothommo committed:
Fix code blocks and signatures in Halcyon

bennothommo committed:
Fix default config namespaces not loading

bennothommo committed:
Further fixes to docblocks and signatures

bennothommo committed:
Documenting magic methods and properties of Auth models

bennothommo committed:
Initial work on strict typing and return types

LukeTowers commented on src/Database/Behaviors/Purgeable.php:
```diff
@@ -4,10 +4,14 @@ class Purgeable extends \Winter\Storm\Extension\ExtensionBase
-     *
-     * public $purgeable = [];
      */
+    public $purgeable = [];
```
Why was this changed?

bennothommo replied:
It was being reported as a missing variable otherwise, but yeah I realise this is probably incorrect since purgeable needs to exist on the model. Will adjust.

LukeTowers commented on composer.json:
```diff
@@ -52,7 +52,9 @@
-        "dms/phpunit-arraysubset-asserts": "^0.1.0|^0.2.1"
+        "dms/phpunit-arraysubset-asserts": "^0.1.0|^0.2.1",
+        "nunomaduro/larastan": "^1.0",
+        "orchestra/testbench": "dev-master"
```
Why was testbench added here?

bennothommo replied:
It's needed for testing Laravel modules and not "true" applications - which Storm kinda sorta is not. I found I needed it to even get PHPStan working in the first place.

LukeTowers commented on phpstan.neon:
```diff
@@ -0,0 +1,9 @@
+        - src
+    level: 5
+    excludePaths:
+        - src/Assetic/*
```
We need to get rid of the custom usage of Assetic in the first place

bennothommo replied:
Indeed, but this is in there for now until that happens.

LukeTowers replied:
@bennothommo we can now get rid of this, although I'm not opposed to integrating Assetic with phpstan as well on its own repo.

LukeTowers commented on src/Halcyon/Datasource/Datasource.php:
```diff
@@ -1,56 +1,91 @@
+    public function makeCacheKey(string $name = ''): string
     {
-        return crc32($name);
+        return hash('crc32b', $name);
```
What's the reasoning behind changing the hash algorithm?

bennothommo replied:
It was the recommendation from the [`crc32()` method docs](https://www.php.net/crc32) if you needed hashes as strings.

LukeTowers commented on src/Halcyon/Datasource/DbDatasource.php:
```diff
@@ -90,27 +86,22 @@ public function getQuery($ignoreDeleted = true)
 
     /**
-     * Helper to make file path.
+     * Helper method to make the full file path to the model.
```
"the full" implies the absolute path to the file on the filesystem, which is incorrect for databases and may not be correct for the filesystem source either (haven't looked). Thoughts?

bennothommo replied:
Perhaps "Helper method to make the full relative file path to the model"? I just thought the original comment was not descriptive enough.

LukeTowers commented on src/Halcyon/Datasource/DbDatasource.php:
```diff
@@ -372,52 +337,31 @@ public function delete(string $dirName, string $fileName, string $extension)
-    public function makeCacheKey($name = '')
-    {
-        return crc32($this->source . $name);
-    }
```
Why was this method removed?

bennothommo replied:
The `makeCacheKey` method is being provided by the abstract class `Winter\Storm\Halycon\Datasource\Datasource`. While this implementation of the method does use `$this->source`, the `getPathsCacheKey()` method has the source as part of the cache key prefix anyway, so it should still be unique. Is there any other reason why the source was necessary here?

LukeTowers commented on src/Halcyon/Datasource/FileDatasource.php:
```diff
@@ -348,54 +299,38 @@ protected function makeDirectoryPath($dirName, $relativePath = '')
-    public function makeCacheKey($name = '')
-    {
-        return crc32($this->basePath . $name);
-    }
```
Why was this method removed?

bennothommo replied:
Same reason as this: https://github.com/wintercms/storm/pull/64/files/c99be36471e01ac8ef93e5c0258a514ac850b866#r771805382

LukeTowers commented on src/Halcyon/Model.php:
```diff
@@ -681,7 +682,8 @@ public function getAttribute($key)
          */
-        if (($attr = $this->fireEvent('model.beforeGetAttribute', [$key], true)) !== null) {
+        $attr = $this->fireEvent('model.beforeGetAttribute', [$key], true);
+        if ($attr !== null) {
```
```suggestion
        if (!is_null($attr)) {
```

LukeTowers commented on src/Halcyon/Model.php:
```diff
@@ -697,7 +699,8 @@ public function getAttribute($key)
          */
-        if (($_attr = $this->fireEvent('model.getAttribute', [$key, $value], true)) !== null) {
+        $_attr = $this->fireEvent('model.getAttribute', [$key, $value], true);
+        if ($_attr !== null) {
```
```suggestion
        if (!is_null($_attr)) {
```

LukeTowers commented on src/Halcyon/Model.php:
```diff
@@ -980,12 +983,12 @@ public function delete()
     protected function performDeleteOnModel()
     {
-        $this->newQuery()->delete($this->fileName);
+        $this->newQuery()->delete();
```
Why was this changed?

bennothommo replied:
The delete method on the model doesn't accept any parameters.

LukeTowers commented on src/Halcyon/Model.php:
```diff
@@ -1553,9 +1557,13 @@ public static function initCacheItem(&$item)
     {
-        if (MemoryCacheManager::isEnabled() && self::getCacheManager() !== null) {
-            self::getCacheManager()->driver()->flushInternalCache();
+        if (!MemoryCacheManager::isEnabled() || self::getCacheManager() === null) {
```
```suggestion
        if (!MemoryCacheManager::isEnabled() || is_null(self::getCacheManager())) {
```

LukeTowers commented on src/Halcyon/Processors/SectionParser.php:
```diff
@@ -182,9 +184,9 @@ public static function parseOffset($content)
+     * @return int|null The line number the instance was found.
      */
-    private static function calculateLinePosition($content, $instance = 1)
+    public static function calculateLinePosition($content, $instance = 1)
```
Why was this changed?

bennothommo replied:
PHPStan reports this as unsafe private static call, since it's being called statically within the class. Realistically, I think this should probably be changed to a singleton.

LukeTowers commented on src/Halcyon/Processors/SectionParser.php:
```diff
@@ -209,7 +211,7 @@ private static function calculateLinePosition($content, $instance = 1)
      * @return int The adjusted line number.
      */
-    private static function adjustLinePosition($content, $startLine = -1)
+    public static function adjustLinePosition($content, $startLine = -1)
```
Why was this changed?

bennothommo replied:
Same as above.

LukeTowers commented on src/Support/Traits/Emitter.php:
```diff
@@ -139,7 +139,8 @@ public function unbindEvent($event = null)
      * @param array $params Event parameters
      * @param boolean $halt Halt after first non-null result
-     * @return array Collection of event results / Or single result (if halted)
+     * @return array|mixed|null Collection of event results as an array if not halted, if not halted. Otherwise, the first non-null result, or null
+     *  if no listeners returned a result.
```
```suggestion
     * @return array|mixed|null If halted, the first non-null result. If not halted, an array of event results. Returns 
     *  null if no listeners returned a result.
```

bennothommo committed:
Apply suggestions from code review

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Merge remote-tracking branch 'origin/wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Set strict types on Argon

bennothommo committed:
Fix comment on makeFilePath method in DbDatasource

bennothommo committed:
Use proper exception for insert, update docs

bennothommo committed:
Fix code smell

bennothommo committed:
Model/Pivot constructor changes.

- Enforce only the "$attributes" parameter to be available to the constructor via an Interface.
- Build Pivot models using a static "fromAttributes" / "fromRawAttributes" method as opposed to a custom constructor.

bennothommo committed:
Adjust users table migration

This should hint that extensions to the User framework should have an "is_superuser" field available for super-user functionality that is expected in Winter.

bennothommo committed:
Improve docblocks in Auth module

bennothommo committed:
Add database migration paths for PHPStan

bennothommo committed:
Extend Laravel Config repo in our own repo

- Also added the "getMany" method that Laravel now contains.

While we have customised our own Config repository implementation, we still follow Laravel in many aspects, so we should extend their class to allow classes which resolve to Laravel's Config repository directly (as opposed to the contract) to still work.

bennothommo committed:
Ignore Auth Manager in analysis

Our Auth Manager implementation clashes a lot with Illuminate's Authenticable implementation.

We'll need to revisit this at some point.

bennothommo committed:
Re-add ApplicationException use case

bennothommo committed:
Inherit docs

bennothommo committed:
Fix return types in dispatcher

bennothommo committed:
Fix docblocks and signatures in Exception module

bennothommo committed:
Fix tests

bennothommo committed:
Adjust docs and signatures for Extendable trait

bennothommo committed:
Return null if no default datasource is set

bennothommo committed:
Fix signatures and docblocks of Filesystem module

bennothommo committed:
Fix Foundation module docblocks and signatures

bennothommo committed:
WIP FormBuilder

bennothommo committed:
WIP Html module analysis

bennothommo committed:
Merge branch 'wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Update Composer deps, small tweaks

bennothommo committed:
Add class comment

bennothommo committed:
Fix Stan issues in Translator

bennothommo committed:
Fix Validation and Support PHPStan issues

bennothommo committed:
Fix Router PHPStan issues

bennothommo committed:
Merge branch 'wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Add Event facade methods

bennothommo committed:
Fix Parse PHPStan issues

bennothommo committed:
Fix Network PHPStan issues

bennothommo committed:
Merge branch 'wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Fix return type of fired events

bennothommo committed:
Fix minor issue in MailManager

bennothommo committed:
WIP Mailer fixes

bennothommo committed:
Fix HTML PHPStand issues

bennothommo committed:
Fix unit tests - Mail ones are not passing now, but pass on main branch

LukeTowers commented on src/Config/LoaderInterface.php:
```diff
@@ -31,6 +31,15 @@ public function exists($group, $namespace = null);
+     * @param string $alias
+     * @return void
+     */
+    public function registerNamespaceAlias(string $namespace, string $alias);
```
Why was this required?

bennothommo replied:
I felt it made sense to have it - not that it's likely, but if anyone creates another type of Config loader, they'll need to support aliases.

LukeTowers commented on src/Config/Repository.php:
```diff
@@ -115,6 +116,27 @@ public function get($key, $default = null)
+        }
+
+        return $config;
+    }
```
Why was this added?

bennothommo replied:
The `Repository` class now extends Laravel's `Repository` class, but to support aliases, we needed to adjust `getMany` to query through the `get` method - Laravel's original `getMany` just simply extracts from an array.

LukeTowers commented on src/Config/Repository.php:
```diff
@@ -148,48 +170,6 @@ public function set($key, $value = null)
-     *
-     * @param  string  $key
-     * @param  mixed  $value
-     * @return void
-     */
-    public function prepend($key, $value)
-    {
-        $array = $this->get($key);
-
-        array_unshift($array, $value);
-
-        $this->set($key, $array);
-    }
-
-    /**
-     * Push a value onto an array configuration value.
-     *
-     * @param  string  $key
-     * @param  mixed  $value
-     * @return void
-     */
-    public function push($key, $value)
-    {
-        $array = $this->get($key);
-
-        $array[] = $value;
-
-        $this->set($key, $array);
-    }
-
-    /**
-     * Get all of the configuration items for the application.
-     *
-     * @return array
-     */
-    public function all()
-    {
-        return $this->items;
-    }
-
```
Why were these removed?

bennothommo replied:
As above, now that we extend Laravel's base `Repository` class, there's no need for these methods to be overridden.

LukeTowers commented on src/Database/ModelInterface.php:
```diff
@@ -0,0 +1,20 @@
+ *
+ * @author Winter CMS
+ */
+interface ModelInterface
```
Why was this added?

bennothommo replied:
It's in my notes in the PR, but basically, we are now enforcing a signature for the Model constructor that matches Laravel and prevent unsafe static calls. It's something that needs to be discussed, because it could *potentially* be a breaking change, but personally, I don't see a need for extra parameters in the constructor.

LukeTowers commented on src/Extension/ExtendableTrait.php:
```diff
@@ -214,13 +214,16 @@ public function addDynamicProperty($dynamicName, $value = null)
      */
     public function extendClassWith($extensionName)
     {
         if (!strlen($extensionName)) {
```
This should probably be changed to a `!empty()` check and the typehints moved into the method signature directly.

bennothommo replied:
After I've gone through all the PHPStan issues, the next step was to go through and convert the PHP docs to true typehints, so I do intend to do this at some point.

I'll change the `strlen` to `empty` though.

LukeTowers commented on src/Filesystem/Definitions.php:
```diff
@@ -12,22 +12,27 @@
 class Definitions
 {
 
+    /**
+     * Lock down the constructor for this class.
+     */
+    final public function __construct()
+    {
+    }
```
Why is this required?

bennothommo replied:
Because `Definitions` is called statically, this ensures that static calls are safe even with extensions to the class.

LukeTowers replied:
Can you give me an example of unsafe behaviour that this protects against? Apologies if you already have, probably just forgot.

bennothommo replied:
It probably will never be a problem for this class in particular, but since this change has been done to a few classes, I'll try and explain the situation better.

So with classes where we have static calls that create a new instance of the same class, like the `get` method in this class:

```php
    public static function get(string $type): array
    {
        return (new static)->getDefinitions($type);
    }
```

This method will break if someone were to extend the class and add an additional parameter to the constructor, because, as you see above, the method is returning a new instance of the class *without* any parameters. Basically, our static calls are expecting a certain signature for the constructor, which an extending class might break.

Making a constructor final prevents this, by enforcing a certain signature for the constructor, ensuring that the `new static` calls will always work.

LukeTowers commented on src/Filesystem/Definitions.php:
```diff
@@ -37,13 +42,13 @@ public function getDefinitions($type)
-     * and ignorePatterns definitions.
+     * Determines if a path should be ignored based on the ignoreFiles and ignorePatterns definitions.
+     *
+     * Returns `true` if the path is visible, `false` otherwise.
```
This method name and comment is throwing me for a loop, they appear to be saying opposite things?

bennothommo replied:
Your right - the comment is wrong. Will adjust :)

LukeTowers commented on src/Filesystem/Filesystem.php:
```diff
@@ -141,34 +144,33 @@ public function isLocalPath($path, $realpath = true)
      *
-     * @param Illuminate\Filesystem\FilesystemAdapter $disk
+     * @param \Illuminate\Filesystem\FilesystemAdapter $disk
      * @return boolean
```
I don't see a reason to include docblock typehinting when we have method signature type hints setup; especially when the docblock ones don't add any additional information.

bennothommo replied:
I'll leave the docblocks if they add additional context, but yeah, we can probably remove them if the typehints / return type are defined.

LukeTowers commented on src/Filesystem/Zip.php:
```diff
@@ -50,22 +50,23 @@
+     * Lock down the constructor for this class.
      */
-    public static function extract($source, $destination, $options = [])
+    final public function __construct()
```
Why is this required?

bennothommo replied:
Protects against unsafe static calls.

bennothommo committed:
Implement changes from Luke's review

bennothommo committed:
Finalise Mail PHPStan fixes

bennothommo committed:
Merge remote-tracking branch 'origin/wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Fix Command PHPStan issues

bennothommo committed:
Merge remote-tracking branch 'origin/wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Tweak extendable trait

bennothommo committed:
WIP fixes of Database PHPStan issues

LukeTowers commented on src/Console/Command.php:
```diff
@@ -76,7 +76,7 @@ public function complete(CompletionInput $input, CompletionSuggestions $suggesti
                     $suggestionType = 'Values';
                     break;
-                case 'options':
+                default:
```
Why would we want to do this? An invalid case as written right now isn't possible; and if there was one I would rather it crashes than it continues silently. I didn't provide a default case on purpose 😉

bennothommo replied:
Your logic makes PHPStan cry. It complains that the `$dataType` and `$suggestionType` vars *may* not exist.

bennothommo replied:
@LukeTowers how about this instead? https://github.com/wintercms/storm/commit/35b36b7d686d82bc0eeead5fee1d8f3ee55bbb21

Keeps your `switch` call as is, and appeases PHPStan.

LukeTowers replied:
But still fails silently. Could do a default that throws an exception I guess.

bennothommo committed:
Defensive coding tweak

bennothommo committed:
Merge remote-tracking branch 'origin/wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Adjust Command logic path for input type parsing

bennothommo committed:
Fix some issues picked up by PHPStan when global scan is used

LukeTowers commented on src/Halcyon/Processors/SectionParser.php:
```diff
@@ -284,7 +284,7 @@ private static function calculateLinePosition(string $content, int $instance = 1
 
             if ($count === $instance) {
-                return static::adjustLinePosition($content, $number);
+                return self::adjustLinePosition($content, $number);
```
instead of changing this to self the method should be changed to protected and this should remain static. The other private method should be made protected too.

bennothommo committed:
Merge branch 'wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Merge branch 'wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Fix tests

bennothommo committed:
Synchronise docblocks

bennothommo committed:
Drop unnecessary overwrite in hasOne relation.

"getResults" overwrite is already handled correctly in Laravel.

bennothommo committed:
Simplify "key:generate" command overwrite

bennothommo committed:
Tweak checks for URL generator values

bennothommo committed:
Refactor database relations.

- Move all traits into a concerns folder, to keep only supported relations in the base relation folder. Stubs for previously available traits have been kept for BC, but are deprecated.
- Abstract BelongsToMany functionality into a trait to share with this class as well as MorphsToMany.
- Change MorphToMany to extend the base MorphToMany class from Laravel, to maintain parity and covariance.
- Fix multiple PHPStan issues

bennothommo committed:
Make Argon extend Illuminate's Carbon wrapper

bennothommo committed:
Numerous fixes to docblocks and signatures

bennothommo committed:
Exclude some errors that are not helping

bennothommo committed:
Final pass of code analysis tweaks

bennothommo committed:
Fix sortable constant reference and test

bennothommo commented:
I believe I have done all that I can with the code analysis. I've gotten the amount of errors from 1100 to ~130 unique errors, with nearly all of the remaining ones being caused by Eloquent magic that we really can't do much about.

I'm hesitant to introduce this as a test - I think if anything we can treat this as a cleanup and making us more in sync with Laravel. If we wanted to make this a test, we can set a baseline in PHPStan, which basically ignores all remaining errors and will only report on any *new* errors going forward.

At this stage, it's ready to review. I've tested this with the main repo's tests and it's not reporting any errors there, but I haven't done a heap of testing directly in Winter so we'll probably need to have it running for a bit on a real project to ensure everything works as is.

bennothommo committed:
Convert private methods to protected in Section Parser

LukeTowers commented on src/Argon/Argon.php:
```diff
@@ -1,11 +1,12 @@
 <?php namespace Winter\Storm\Argon;
 
 use Carbon\Carbon as DateBase;
```
On line 42 wouldn't it be more accurate to state that the return type is Argon, not DataBase? We could then get rid of the DateBase import.

bennothommo committed:
Change return type in Argon

bennothommo committed:
Synchronise Pivot model with Laravel 9

bennothommo committed:
Add MorphPivot support

LukeTowers commented:
@bennothommo great, looking forward to reviewing this! Before I do, are there any of the breaking changes that we can avoid making? If the breaking changes are just being made to satisfy the tool's need for pedantics then perhaps we can just ignore them as the baseline. I only want this PR to introduce breaking changes if they are absolutely essential and should have been made anyways.

LukeTowers commented on src/Database/Attach/File.php:
```diff
@@ -93,14 +98,10 @@ class File extends Model
-        if ($uploadedFile === null) {
-            return;
-        }
-
```
What was the reason for this change?

bennothommo replied:
`$uploadedFile` should not be null, according to both the docblock and the signature. If it is, this code block is silently discarding it, when it should throw an exception.

LukeTowers commented on src/Database/Attach/File.php:
```diff
@@ -141,17 +142,13 @@ public function fromFile($filePath)
-        if ($data === null) {
-            return;
-        }
-
```
What was the reason for this change?

bennothommo replied:
Same reason as the `fromPost` changes.

LukeTowers commented on src/Database/Attach/File.php:
```diff
@@ -786,10 +794,9 @@ protected function putFile($sourcePath, $destinationFileName = null)
             !FileHelper::isDirectory($destinationPath) &&
-            !FileHelper::makeDirectory($destinationPath, 0777, true, true) &&
-            !FileHelper::isDirectory($destinationPath)
+            !FileHelper::makeDirectory($destinationPath, 0777, true, true)
```
What was the reason for this change?

bennothommo replied:
The second check to `isDirectory` is redundant. If `makeDirectory` fails, it's already returning `false` which triggers the error.

LukeTowers commented on src/Database/Attach/File.php:
```diff
@@ -866,14 +873,13 @@ protected function deleteEmptyDirectory($dir = null)
-        if (!$dir) {
-            return null;
-        }
-
```
What was the reason for this change? It conflicts with the docblock return type now

bennothommo replied:
You're right. Fixed the docblock.

LukeTowers commented on src/Database/Behaviors/Purgeable.php:
```diff
@@ -1,13 +1,16 @@
 
+/**
+ * @deprecated 1.2.0. We recommend using the \Winter\Storm\Database\Traits\Purgeable trait instead.
+ */
```
Why?

bennothommo replied:
I'm not really sure why anymore. I think it was when I was railing against behaviours, but you're right, they have their use. We should probably make this share functionality with the trait if possible though, so we don't have to modify code in two places.

LukeTowers replied:
I think we tried when we first added it and it wasn't that easy to do.

LukeTowers commented on src/Database/Behaviors/Sortable.php:
```diff
@@ -23,6 +23,7 @@
  *
  *   const SORT_ORDER = 'my_sort_order';
  *
+ * @deprecated 1.2.0. We recommend using the \Winter\Storm\Database\Traits\Sortable trait instead.
```
Why? The DB traits as behaviours are intended for use by plugins extending other plugins.

mjauvin replied:
Yeah, I added this to dynamically make a model Sortable when extending a plugin.

LukeTowers commented on src/Database/Connections/Connection.php:
```diff
@@ -50,14 +48,31 @@ public function logQuery($query, $bindings, $time = null)
+     * @param array|object $attributes
+     * @return void
+     */
+    protected function fireEvent($event, $attributes = [])
```
Is this a Laravel override? If not we can probably add the type hints direct to the signature instead of the docblock

bennothommo replied:
It's not, no. I'll add the type hints.

LukeTowers commented on src/Database/MorphPivot.php:
```diff
@@ -0,0 +1,185 @@
+
+        return $query;
+    }
+}
```
Do we have WIP documentation for this?

bennothommo replied:
Not yet, I still need to add some unit tests and make sure it works.

LukeTowers replied:
Any progress on docs and unit tests? Can we copy the DB test from Laravel for this?

bennothommo replied:
Laravel had barely any unit tests for MorphPivot in particular. I've added a couple.

LukeTowers commented on src/Database/Traits/Purgeable.php:
```diff
@@ -54,26 +54,24 @@ public function addPurgeable($attributes = null)
         $originalAttributes = array_diff_key($attributes, $cleanAttributes);
 
-        if (is_array($this->originalPurgeableValues)) {
+        if (count($this->originalPurgeableValues)) {
```
Can this property ever be anything that isn't countable()? Why was this change needed?

bennothommo replied:
No, it'll always be an array.

LukeTowers commented on src/Database/Traits/Sortable.php:
```diff
@@ -75,6 +75,6 @@ public function setSortableOrder($itemIds, $itemOrders = null)
     public function getSortOrderColumn()
     {
-        return defined('static::SORT_ORDER') ? static::SORT_ORDER : 'sort_order';
+        return defined('static::SORT_ORDER') ? constant('static::SORT_ORDER') : 'sort_order';
```
Why?

bennothommo replied:
Both PHPStan AND Intellisense/Intelephense report it as a missing constant, even with the `defined` check surrounding it.

LukeTowers replied:
Silly automated tools...

LukeTowers commented on src/Foundation/Exception/Handler.php:
```diff
@@ -176,14 +176,14 @@ protected function callCustomHandlers($throwable, $fromConsole = false)
             // get sent back to the browsers. Once the handler returns a valid response
             // we will cease iterating through them and calling these other handlers.
-            if (isset($response) && ! is_null($response)) {
+            if (isset($response)) {
```
Why was this required?

bennothommo replied:
`isset` and `!is_null` is the same test.

LukeTowers reviewed (commented):
Some initial comments before I head to bed, looking good so far @bennothommo!

bennothommo commented:
@LukeTowers I feel all of the BC breaks above will be necessary, as they either fix broken functionality, move us closer to Laravel or are documentation changes. There might be some small tweaks outside of those that may not be necessary and were just to appease PHPStan, but at this point, I think it all is in service of making sure our documentation is accurate and our functionality works smoothly. The idea of setting the baseline is that we have as few issues as possible, because otherwise there's a lot more chance for errors to be reported in the future.

bennothommo commented on src/Database/Attach/File.php:
```diff
@@ -866,14 +873,13 @@ protected function deleteEmptyDirectory($dir = null)
+     *
+     * @param string $dir The path to the directory.
+     *
+     * @return bool|null
```
```suggestion
     * @return bool
```

bennothommo committed:
Update src/Database/Attach/File.php

LukeTowers commented on src/Console/Command.php:
```diff
@@ -97,6 +97,8 @@ public function complete(CompletionInput $input, CompletionSuggestions $suggesti
                     $suggestionType = 'Options';
                     break;
+                default:
+                    throw new \Exception('Invalid input type being parsed during completion');
```
```suggestion
                    // This should not be possible to ever be triggered given the type is hardcoded above
                    throw new \Exception('Invalid input type being parsed during completion');
```

LukeTowers commented on src/Database/Attach/File.php:
```diff
@@ -93,14 +98,10 @@ class File extends Model
-     * @param Symfony\Component\HttpFoundation\File\UploadedFile $uploadedFile
+     * @param \Symfony\Component\HttpFoundation\File\UploadedFile $uploadedFile
      */
     public function fromPost($uploadedFile)
```
Should we type hint the parameter now to make the error more explicit when incorrect data is provided rather than being "Failed to access method getClientOriginalName() on null"?

LukeTowers commented on src/Database/Attach/File.php:
```diff
@@ -141,17 +142,13 @@ public function fromFile($filePath)
     {
-        if ($data === null) {
-            return;
-        }
```
Likewise, should we add the type hint to the method signature now?

LukeTowers commented on src/Database/Model.php:
```diff
@@ -529,7 +529,7 @@ public function getObservableEvents()
      * Get a fresh timestamp for the model.
      *
-     * @return \Winter\Storm\Argon\Argon
+     * @return \Illuminate\Support\Carbon
```
Why change this? Isn't Argon more specific than Carbon?

bennothommo replied:
It matches Laravel's signature, I believe.

LukeTowers commented on src/Database/Models/Revision.php:
```diff
@@ -34,7 +34,7 @@ public function getNewValueAttribute($value)
     public function getOldValueAttribute($value)
     {
-        if ($this->cast == 'date' && !is_null($value)) {
+        if ($this->getAttribute('cast') === 'date' && !is_null($value)) {
```
Why was this required?

bennothommo replied:
It fixes an issue where the `cast` property might not be set.

LukeTowers replied:
Wouldn't that just cycle through the getAttribute() logic anyways by way of the `__get()` magic method?

bennothommo replied:
It does, but where possible, we should avoid relying on magic in internal methods as it's not easily deducible. Userland code may use magic.

LukeTowers replied:
Hmm, I'm not sure. It would change the behaviour if someone had specified a `cast` property on the model - in these sorts of cases I feel like it's better to use the standard way of interacting with model attributes rather than bypass one of the layers.

bennothommo replied:
Sorry @LukeTowers, just circling back to this only now. I'm not sure what you mean here though - both using this `getAttribute` call or using the `__get` magic method do the same thing, so what behaviour would be changed here?

LukeTowers replied:
@bennothommo I mean that if someone defines `$cast` as a property on a class that extends this model class then this logic would no longer get that property value. However, this is unlikely so I no longer care about this change :P

LukeTowers commented on src/Database/MorphPivot.php:
```diff
@@ -0,0 +1,185 @@
+<?php namespace Winter\Storm\Database;
+
+class MorphPivot extends Pivot
```
```suggestion
/**
 * Reimplementation of \Illuminate\Database\Eloquent\Relations\MorphPivot
 * but extends Winter's Pivot base class instead of Laravel's to get access to
 * Winter magic.
 */
class MorphPivot extends Pivot
```

LukeTowers commented on src/Database/Pivot.php:
```diff
@@ -1,17 +1,24 @@
-     * @var \Illuminate\Database\Eloquent\Model
+     * @var \Illuminate\Database\Eloquent\Model|null
      */
     protected $parent;
```
Why was this property retained instead of replaced as I did in https://github.com/wintercms/storm/pull/65?

bennothommo replied:
I couldn't discount that `$parent` might be used by our users. The edits I made to the Pivot keep it backwards compatible.

LukeTowers replied:
@bennothommo we should be able to use the AsPivot trait instead in our model and just handle the BC for the parent property by defining getParentAttribute() & setParentAttribute() methods. Thoughts?

LukeTowers commented on src/Database/Pivot.php:
```diff
@@ -1,17 +1,24 @@
-use Illuminate\Database\Eloquent\Builder as BuilderBase;
+use Winter\Storm\Support\Str;
 
 class Pivot extends Model
```
Instead of replicating a the AsPivot trait in our code, can we just implement it in our Pivot model and then override any methods that need tweaking?

https://github.com/laravel/framework/blob/9.x/src/Illuminate/Database/Eloquent/Relations/Concerns/AsPivot.php

bennothommo committed:
Update src/Console/Command.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/MorphPivot.php

Co-authored-by: Luke Towers <github@luketowers.ca>

LukeTowers commented on src/Database/MorphPivot.php:
```diff
@@ -0,0 +1,190 @@
+ * but extends Winter's Pivot base class instead of Laravel's to get access to
+ * Winter magic.
+ */
+class MorphPivot extends Pivot
```
@bennothommo I've added a [comment](https://github.com/laravel/framework/pull/25851#issuecomment-1133820077) to the original AsPivot PR making a case for the addition of an AsMorphPivot trait, we could potentially submit the PR ourselves to simplify our codebase.

LukeTowers commented on src/Database/QueryBuilder.php:
```diff
@@ -112,14 +113,10 @@ public function get($columns = ['*'])
     {
-        if (is_null($this->columns)) {
-            $this->columns = $columns;
-        }
```
Why was this removed?

bennothommo replied:
I think this was due to PHPStan saying that `$this->columns` can never be null.

LukeTowers commented on src/Database/QueryBuilder.php:
```diff
@@ -140,14 +137,10 @@ protected function getDuplicateCached($columns = ['*'])
     {
-        if (is_null($this->columns)) {
-            $this->columns = $columns;
-        }
```
Why was this removed?

bennothommo replied:
Same answer as https://github.com/wintercms/storm/pull/64#discussion_r878865484

LukeTowers commented on src/Database/Relations/AttachMany.php:
```diff
@@ -81,7 +80,7 @@ public function getSimpleValue()
         if ($files) {
             $value = [];
-            foreach ($value as $file) {
+            foreach ($files as $file) {
```
Does this cause any side effects having this unreported bug fixed?

bennothommo replied:
Perhaps it would make attachments be displayed correctly in the Lists widget? That's it? :P

LukeTowers commented on src/Database/Behaviors/Purgeable.php:
```diff
@@ -1,13 +1,16 @@
 
+/**
+ * @deprecated 1.2.0. We recommend using the \Winter\Storm\Database\Traits\Purgeable trait instead.
+ */
```
```suggestion
```

LukeTowers commented on src/Database/Behaviors/Sortable.php:
```diff
@@ -23,6 +23,7 @@
  *
  *   const SORT_ORDER = 'my_sort_order';
  *
+ * @deprecated 1.2.0. We recommend using the \Winter\Storm\Database\Traits\Sortable trait instead.
```
```suggestion
```

LukeTowers commented on src/Database/Relations/AttachOneOrMany.php:
```diff
@@ -1,303 +1,9 @@
-use Symfony\Component\HttpFoundation\File\UploadedFile;
-
+/**
+ * @deprecated 1.2.0 Use `\Winter\Storm\Database\Relations\Concerns\AttachOneOrMany` instead.
```
Can you explain more as to what's the motivation behind moving all these relation classes to a new location? What's the benefit?

bennothommo replied:
Due to the naming of the classes, it was getting a bit too hard to track what was an actual relation vs. what was a trait when all these files were mixed in the same directory under `src/Database/Relations`. I thought a tidy up would be useful here, in making sure that the traits were in a concerns subdirectory, and only the relations were kept in the main directory.

At the same time, I realised people could've been using the traits directly, so I had to keep the stubs in place. I would say though that it would be beneficial to remove them in the next main release.

TL;DR: Just clean-up of the codebase.

LukeTowers replied:
How likely is it that people were actually using these traits? What would be the use case for that? If we're doing it to cleanup and it's not that likely that they were actually being used then we might as well just remove them in this version, otherwise we're not really seeing the benefits of the cleanup.

bennothommo replied:
I don't think it would be very likely anyone was using the traits directly.

LukeTowers replied:
Let's just document the change and remove the original traits then.

LukeTowers commented on src/Extension/ExtendableTrait.php:
```diff
@@ -1,12 +1,12 @@
 use BadMethodCallException;
 use Winter\Storm\Support\ClassLoader;
 use Winter\Storm\Support\Serialization;
+use Illuminate\Support\Facades\App;
```
Why do some files need to have the short form import removed but others don't? What's the point?

bennothommo replied:
Are you meaning the full facade names? I'm replacing the short facade names or inline uses with the full facade names everywhere I see.

LukeTowers replied:
Yes, what's the benefit in doing that?

bennothommo replied:
I think we alias the short names in the System module, not the library, so when running static analysis or doing code edits directly on the library, it isn't able to determine the correct class from the short names, so this change makes it explicit.

LukeTowers replied:
Gotcha. Can we do some find and replace to make it fully explicit everywhere in this repo then?

LukeTowers commented on src/Filesystem/Filesystem.php:
```diff
@@ -141,34 +144,30 @@ public function isLocalPath($path, $realpath = true)
+     * Determines if a file exists (ignoring the case for the filename only).
      */
-    public function existsInsensitive($path)
+    public function existsInsensitive(string $path): string|false
```
Is `false` a valid return type in PHP 8?

bennothommo replied:
It is indeed.

![image](https://user-images.githubusercontent.com/15900351/169698823-4a378aeb-d942-42aa-908f-af335d3fee2b.png)

LukeTowers commented on src/Filesystem/Filesystem.php:
```diff
@@ -141,34 +144,30 @@ public function isLocalPath($path, $realpath = true)
+     * @param string|object $className Class name or object
      */
-    public function fromClass($className)
+    public function fromClass(string|object $className): string|false
```
Is `false` a valid return type in PHP 8?

bennothommo replied:
Same answer as https://github.com/wintercms/storm/pull/64#discussion_r878866343

LukeTowers commented on src/Filesystem/Filesystem.php:
```diff
@@ -191,52 +190,48 @@ public function existsInsensitive($path)
-        if (!$firstChar = $this->isPathSymbol($path)) {
-            return $default === false ? $path : $default;
+        if (!$this->isPathSymbol($path)) {
+            return (is_null($default)) ? $path : $default;
```
Why is this required? Wouldn't this mean that the return type cannot be only string since passing a path with no path symbol and a `$default` value of `false` or `true` will result in a bool being returned instead of a string? How come static analysis didn't catch this?

bennothommo replied:
Not sure why it wasn't picked up? Maybe a higher level check?

LukeTowers commented on src/Filesystem/Filesystem.php:
```diff
@@ -191,52 +190,48 @@ public function existsInsensitive($path)
     {
-        $firstChar = substr($path, 0, 1);
-        if (isset($this->pathSymbols[$firstChar])) {
-            return $firstChar;
```
Any concerns about this breaking change?

bennothommo replied:
I'll make a note of it, but given the docblock itself said it was *supposed* to be a boolean, and the name of the function implies a boolean check, I think it's very unlikely people are using it to extract the path symbol.

LukeTowers commented on src/Filesystem/Filesystem.php:
```diff
@@ -411,10 +405,8 @@ public function fileNameMatch($fileName, $pattern)
 
     /**
      * Finds symlinks within the base path and provides a source => target array of symlinks.
```
```suggestion
     * Finds symlinks within the base path and populates the local symlinks property with an array of source => target symlinks.
```

LukeTowers commented on src/Filesystem/Zip.php:
```diff
@@ -50,22 +50,23 @@
+    /**
+     * Extracts an existing ZIP file.
+     */
+    public static function extract(string $source, string $destination, array $options = []): bool
```
Can we document the valid elements for the `$options` array? Looks like right now all that's supported is the `mask` element.

LukeTowers commented on src/Filesystem/Zip.php:
```diff
@@ -108,13 +105,13 @@ public static function make($destination, $source, $options = [])
+     * Adds a source file or directory to a Zip file.
      */
-    public function add($source, $options = [])
+    public function add(string $source, array $options = []): self
```
Can we document the valid options here as well?

LukeTowers commented on src/Filesystem/Zip.php:
```diff
@@ -124,23 +121,18 @@ public function add($source, $options = [])
-            'baseglob' => basename($source)
-        ], $options));
+        $basedir = dirname($source);
+        $baseglob = basename($source);
```
Isn't this a breaking change by no longer supporting `$basedir` and `$baseglob` in the options array?

bennothommo replied:
Re-added the ability to overwrite these two variables through the options.

LukeTowers commented on src/Foundation/Application.php:
```diff
@@ -390,12 +390,12 @@ public function setLocale($locale)
     public function registerConfiguredProviders($isRetry = false)
     {
-        $providers = Collection::make($this->config['app.providers'])
+        $providers = Collection::make($this->get('config')['app.providers'])
```
Why is this required?

bennothommo replied:
Because PHPStan is unable to follow through the magic methods, like with the example before with model attributes. This makes it explicit what is happening in a way that PHPStan can follow.

LukeTowers commented on src/Halcyon/ModelInterface.php:
```diff
@@ -0,0 +1,20 @@
+ *
+ * @author Winter CMS
+ */
+interface ModelInterface
```
Why wouldn't Halcyon use the Winter Eloquent model interface?

bennothommo replied:
Just seemed to be cleaner to have its own interface - having it implement the database interface may be misconstrued as Halcyon operating off the databases.

LukeTowers commented on src/Halcyon/Processors/SectionParser.php:
```diff
@@ -65,13 +65,14 @@ protected static function parseIntoSections(string $content, int $limit = 3): ar
-        if (!$isCompoundObject) {
-            return array_get($data, 'content', '');
+        if (!isset($isCompoundObject) || $isCompoundObject === false) {
+            return array_get($data, 'content');
```
Why did you remove the empty string as the default value here?

bennothommo replied:
Not sure, actually. I'll re-add it.

LukeTowers commented on src/Html/BlockBuilder.php:
```diff
@@ -75,30 +75,21 @@ public function endBlock($append = false)
+    public function append(string $name, string $content): void
     {
         if (!isset($this->blocks[$name])) {
             $this->blocks[$name] = '';
```
Unrelated to the current PR, doesn't this mean that if you call append targeting a block that doesn't exist yet if you call it again a second time it will exist in the second instance but not the first?

bennothommo replied:
Correct - if you try to append a block that doesn't exist, this acts more or less like the `set` method. The second time you run it, the block would exist and this would therefore append to the block.

LukeTowers commented on src/Parse/Markdown.php:
```diff
@@ -90,16 +105,19 @@ public function parseLine($text)
-        $this->fireEvent('beforeParse', $data, false);
-        Event::fire('markdown.beforeParse', $data, false);
+        $this->fireEvent('beforeParse', [$data], false);
+        Event::fire('markdown.beforeParse', [$data], false);
```
Does this cause any issues having it changed to an array instead of just the variable?

bennothommo replied:
It is listed as one of the breaking changes. Since event payloads are processed into an array, this was done mainly to appease PHPStan, but it's probably something we should stick to with events anyway.

LukeTowers commented on src/Parse/Markdown.php:
```diff
@@ -31,14 +31,37 @@ class Markdown
+     */
+    protected function getParser()
+    {
+        return new $this->parserClass;
```
Why was this changed to no longer cache the parser class instance?

bennothommo replied:
It was to prevent cross-contamination of data with subsequent Markdown calls. Each Markdown call will now have a clean instance of the parser to work with.

LukeTowers replied:
Does the parser contain dangerous state?

bennothommo replied:
Not confirmed, but it was better to be safe. I don't expect people would use too many Markdown calls in a single request.

LukeTowers commented on src/Support/Facade.php:
```diff
@@ -10,31 +10,4 @@
-    protected static function getFacadeInstance()
-    {
-        return null;
-    }
```
Why were these changes required?

bennothommo replied:
As far as I can tell, this method is not used anywhere in the code. ~~Seemed to be a holdover from a previous Laravel version.~~

EDIT: Actually, I think this was to bring us closer in line to Laravel.

LukeTowers commented on src/Support/Facades/Input.php:
```diff
@@ -18,6 +21,18 @@ public static function get($key = null, $default = null)
+    public static function all()
+    {
+        return static::$app['request']->input();
+    }
```
Why was this required?

bennothommo replied:
Mainly for the `input()` helper method - it will try to retrieve all input if no name is provided to that method. This is synonymous with the `post()` and `get()` helper methods.

LukeTowers commented on src/Support/Serialization.php:
```diff
@@ -12,11 +12,11 @@ class Serialization
     public static function wrapClosure($callable)
     {
-        if ($callable instanceof Closure && !($callable instanceof SerializableClosure)) {
+        if ($callable instanceof Closure) {
```
Won't this now double wrap SerializableClosures?

bennothommo replied:
`SerializableClosure` is not an instance of `Closure`, so it should skip this condition block and simply be returned.

LukeTowers commented on src/Validation/Factory.php:
```diff
@@ -10,6 +10,13 @@
+     *
+     * @var \Closure|null
+     */
+    protected $resolver;
```
Why was this change required?

bennothommo replied:
It was an attempt to appease PHPStan on another line - there is an `is_null` check for `$this->resolver` which PHPStan says will always be true because according to the docs for resolver, it can never be `null`. This is not strictly speaking correct, but it is how the API defines it.

It doesn't work, because now the doc is not covariant with Laravel's original docs, so I'll remove it and rely on Laravel's definition.

bennothommo committed:
Add type hints to all File attachment methods

bennothommo committed:
Add type hints

bennothommo committed:
Make the Pivot use the AsPivot trait, add parent accessor

bennothommo committed:
Update src/Database/Behaviors/Purgeable.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Behaviors/Sortable.php

Co-authored-by: Luke Towers <github@luketowers.ca>

LukeTowers commented on src/Database/Attach/File.php:
```diff
@@ -270,33 +258,34 @@ public function getSizeAttribute()
             'Content-type'        => $this->getContentType(),
             'Content-Disposition' => $disposition . '; filename="' . $this->file_name . '"',
-            'Cache-Control'       => 'private, no-store, no-cache, must-revalidate, pre-check=0, post-check=0, max-age=0',
+            'Cache-Control'       => 'private, no-store, no-cache, must-revalidate, max-age=0',
```
What's the reason for this change?

bennothommo replied:
`pre-check` and `post-check` are invalid Cache Control attributes. Internet Explorer 5 introduced them, and they haven't been used for years. [Relevant blog post for more info](https://docs.microsoft.com/en-us/archive/blogs/ieinternals/internet-explorers-cache-control-extensions?_fsi=f81xrzF2)

LukeTowers commented on src/Database/Attach/File.php:
```diff
@@ -397,15 +385,15 @@ public function getContentType()
      * Returns the public address to access the file.
      */
-    public function getPath($fileName = null)
+    public function getPath(?string $fileName = null): string
```
Adding type hints to all of these methods means that we'll need to propagate them to the System\Models\File model as well as any userland extensions of that class.

LukeTowers replied:
@bennothommo ping

bennothommo replied:
What would you prefer? Roll them back or mark it as a breaking change? I'm erring towards marking it as breaking, since it would be better to have full type hinting, and I can't imagine too many people would've extended this model.

LukeTowers commented on src/Database/Behaviors/Purgeable.php:
```diff
@@ -76,12 +76,7 @@ public function purgeAttributes($attributesToPurge = null)
-        else {
-            $this->originalPurgeableValues = $originalAttributes;
-        }
+        $this->originalPurgeableValues = array_merge($this->originalPurgeableValues, $originalAttributes);
```
Why was this change made? Does this property always exist as at least an array? If so where is it defined?

bennothommo replied:
It will always be an array, so the condition that was used previously is redundant.

Lines 68-70 above will convert a single string provided to the argument into an array, and the `purgeable` property itself is defined as an array.

LukeTowers committed:
Update src/Filesystem/Filesystem.php

bennothommo committed:
Merge branch 'wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Remove relation traits that have been moved to concerns

bennothommo committed:
Update all facade use cases to use full facade class

bennothommo committed:
Add options docs for Zip class, fix BC break

bennothommo committed:
Re-add default value

bennothommo committed:
Remove overwritten definition for resolver

LukeTowers commented on src/Database/Relations/HasOne.php:
```diff
@@ -22,21 +22,6 @@ public function __construct(Builder $query, Model $parent, $foreignKey, $localKe
-        }
-
-        return parent::getResults();
-    }
```
Left over code from a bygone era?

bennothommo replied:
Nah, the parent classes do have that method, and handle it better from what I can see, so just using that.

LukeTowers commented on src/Html/FormBuilder.php:
```diff
@@ -525,48 +458,33 @@ public function selectRange($name, $begin, $end, $selected = null, $options = []
 
         foreach (range(1, 12) as $month) {
-            $months[$month] = strftime($format, mktime(0, 0, 0, $month, 1));
+            $months[$month] = strftime($format, mktime(12, 0, 0, $month, 1));
```
Why was this change needed?

bennothommo replied:
It's a quick fix that I deploy on most sites to ensure that timezones don't affect the result, but thinking about it, it's not really required here, so I'll revert.

LukeTowers commented on src/Html/FormBuilder.php:
```diff
@@ -936,7 +828,7 @@ protected function getAppendage($method)
      * @param  string  $name
      * @param  array   $attributes
-     * @return string
+     * @return string|null
```
Can this return null if you default to returning an empty string?

bennothommo replied:
I think I had interpreted that `null` could be a value within the `$attributes` array, and the `array_key_exists` condition could still pass. But if we go by that logic, then *anything* could be part of that attributes array, so I've changed it to encourage a `string` return.

bennothommo committed:
Merge branch 'wip/1.2' into wip/1.2-code-analysis

bennothommo committed:
Fix moved class names in aliases

bennothommo committed:
Add MorphPivot test cases

bennothommo committed:
Fix constructor for MorphToPivot

bennothommo committed:
Revert hour value change

bennothommo committed:
Correct docblock return

bennothommo committed:
Match PHPCS config from main repo

bennothommo committed:
Use lang path in Translator registration

bennothommo committed:
Fix possible BC break, cleanup for PHPStan

bennothommo committed:
Code tidying

bennothommo committed:
Try out @phpstan-property definitions to fix remaining issues

bennothommo committed:
Docs and commenting changes to fix a whole slew of issues

bennothommo committed:
Generate baseline

bennothommo committed:
Allow pivot model to be provided to MorphTo relations

bennothommo committed:
Fix test, allow namespaced files in tests folder

bennothommo committed:
Move tests namespace into autoload-dev

bennothommo committed:
Add automated PHPStan test, tweak PHPUnit test

bennothommo committed:
Replace missing matrix vars

bennothommo committed:
Fix Markdown parser class name storage and use

bennothommo committed:
Remove use of Eloquent facade in down migrations

bennothommo committed:
Convert all type hints to PHPDocs to maintain BC

bennothommo committed:
Small tweak

bennothommo committed:
Ensure morph to supports deferred records and constraints

bennothommo committed:
Remove fixed errors from baseline

bennothommo committed:
Allow CSRF token and session key to be null

bennothommo committed:
Fix return type

bennothommo committed:
Ignore lines in PHPStan - it can't seem to work out the relations here

LukeTowers commented on src/Database/Model.php:
```diff
@@ -803,15 +805,6 @@ protected function saveInternal($options = [])
-            $this->fireModelEvent('updated', false);
-            $this->fireModelEvent('saved', false);
-        }
-
```
Why was this removed?

bennothommo replied:
It's unreachable code.

LukeTowers commented on src/Database/Relations/Concerns/DeferOneOrMany.php:
```diff
@@ -35,24 +38,27 @@ public function withDeferred($sessionKey)
                             ->from($this->table)
-                            ->where($this->getOtherKey(), DbDongle::raw(DbDongle::getTablePrefix().$this->related->getQualifiedKeyName()))
+                            ->where($this->getOtherKey(), DbDongle::raw(
+                                DbDongle::getTablePrefix().$this->related->getQualifiedKeyName()
```
```suggestion
                                DbDongle::getTablePrefix() . $this->related->getQualifiedKeyName()
```

LukeTowers commented on src/Database/Relations/Concerns/DeferOneOrMany.php:
```diff
@@ -35,24 +38,27 @@ public function withDeferred($sessionKey)
                             ->from($this->table)
-                            ->where($this->getOtherKey(), DbDongle::raw(DbDongle::getTablePrefix().$this->related->getQualifiedKeyName()))
+                            ->where($this->getOtherKey(), DbDongle::raw(
+                                DbDongle::getTablePrefix().$this->related->getQualifiedKeyName()
```
```suggestion
                                DbDongle::getTablePrefix() . $this->related->getQualifiedKeyName()
```

LukeTowers commented on src/Foundation/Providers/ExecutionContextProvider.php:
```diff
@@ -34,14 +33,10 @@ public function register()
         }
 
-        return $url;
+        return (substr($url, 0, 1) !== '/') ? '/' . $url : $url;
```
Should this be Str::startsWith() instead for better readability?

LukeTowers commented on src/Translation/Translator.php:
```diff
@@ -167,9 +167,13 @@ public function parseKey($key)
+        }
 
-        return call_user_func($this->determineLocalesUsing ?: fn () => $locales, $locales);
+        return $locales;
```
Does this mean that the determineLocalesUsing feature doesn't work in Winter?

bennothommo replied:
It's called in the parent class. We're just appending our core language to the end if need be.

bennothommo committed:
Update src/Database/Relations/Concerns/DeferOneOrMany.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Update src/Database/Relations/Concerns/DeferOneOrMany.php

Co-authored-by: Luke Towers <github@luketowers.ca>

bennothommo committed:
Remove return type missed in last pass

bennothommo committed:
Change to Str call
